GoogleUtilities/Tests/Unit/Environment/GULKeychainStorageTests.m

```diff
@@ -146,43 +175,144 @@ - (void)testRemoveNonExistingObject {
+  [self.mockCache removeAllObjects];
+  [self assertSuccessReadObject:@100
+                         forKey:@"test-key1"
+                          class:[NSNumber class]
+                  existsInCache:YES
+                    fromStorage:self.storage
+                  withMockCache:self.mockCache];
+}
+
+- (void)testVersionCompatibility_SetObject {
+  // Given
+  // - Object is set with old implementation.
+  id oldMockCache = OCMPartialMock([[NSCache alloc] init]);
+  GULKeychainStorageV7_7_0 *oldKeychainStorage =
+      [[GULKeychainStorageV7_7_0 alloc] initWithService:kKeychainServiceName cache:oldMockCache];
+
+  [self assertSuccessWriteObject:@100
+                          forKey:@"test-key1"
+                       toStorage:(GULKeychainStorage *)oldKeychainStorage
+                   withMockCache:oldMockCache];
+
+  // When
+  // - App is updated to include a Google Utilities version greater than 7.7.0.
+
+  // Then
+  // - The same object is updated with new implementation.
+  [self assertSuccessWriteObject:@200
+                          forKey:@"test-key1"
+                       toStorage:self.storage
+                   withMockCache:self.mockCache];
+
+  [self assertSuccessReadObject:@200
+                         forKey:@"test-key1"
+                          class:[NSNumber class]
+                  existsInCache:YES
+                    fromStorage:self.storage
+                  withMockCache:self.mockCache];
+
+  [oldMockCache removeAllObjects];
+  [self assertSuccessReadObject:@200
+                         forKey:@"test-key1"
+                          class:[NSNumber class]
+                  existsInCache:NO
+                    fromStorage:(GULKeychainStorage *)oldKeychainStorage
+                  withMockCache:oldMockCache];
+}
+
+- (void)testVersionCompatibility_RemoveObject {
+  // Given
+  // - Object is set with old implementation.
+  [self assertNonExistingObjectForKey:@"test-key1" class:[NSNumber class]];
+
+  id oldMockCache = OCMPartialMock([[NSCache alloc] init]);
+  GULKeychainStorageV7_7_0 *oldKeychainStorage =
+      [[GULKeychainStorageV7_7_0 alloc] initWithService:kKeychainServiceName cache:oldMockCache];
+
+  [self assertSuccessWriteObject:@100
+                          forKey:@"test-key1"
+                       toStorage:(GULKeychainStorage *)oldKeychainStorage
+                   withMockCache:oldMockCache];
+
+  // When
+  // - App is updated to include a Google Utilities version greater than 7.7.0.
+
+  // Then
+  // - The same object is removed with new implementation.
+  [self assertRemoveObjectForKey:@"test-key1"];
+
+  [oldMockCache removeAllObjects];
+  [self assertSuccessReadObject:@100
+                         forKey:@"test-key1"
+                          class:[NSNumber class]
+                  existsInCache:NO
+                    fromStorage:(GULKeychainStorage *)oldKeychainStorage
+                  withMockCache:oldMockCache];
+}
+
```

Comment:
There are three cases to check for version compatibility. The first test is prefixed as fails as it fails locally. The other two cases pass. 

The failure indicates that when apps upgrade to the new implementation, they will not be able to receive any data that was stored with the previous implementation.